src/ledger/LedgerManagerImpl.cpp

```diff
@@ -1028,16 +1025,21 @@ LedgerManagerImpl::closeLedger(LedgerCloseData const& ledgerData)
     //
     // 2. Commit the current transaction.
     //
-    // 3. Start background eviction scan for the next ledger, _after_ the commit
-    //    so that it takes its snapshot of network setting from the
-    //    committed state.
+    // 3. Finalize any new checkpoint files _after_ the commit. If a crash
+    // occurs
+    //   between commit and this step, core will try to rotate files again on
```

Comment:
by rotate, I mean "durably rename dirty checkpoint file to its actual name; the renamed file will be published". Specifically, core calls `CheckpointBuilder::checkpointComplete` method, which will rename the files. An exception is thrown if rename fails. In this case, on restart no prefix recovery is needed since we didn't produce any unwanted data (rename happens _after_ commit, which means checkpoint is ready to be published). So on restart core will again attempt renaming "dirty" files to their proper names. The assumption is that that durable rename should succeed eventually (we do this in the buckets subsystem as well, and it seems to work fine).

Reply:
To clarify, the sequence of events in this failure mode is as-follows:
1. core finishes writing a complete checkpoint, it's still called ".dirty"
2. core queues up checkpoint to publish
3. core commits the ledger, at which point LCL is advanced
4. core finalizes the checkpoint by renaming files to their proper names. 
Any crash before or during step 3 triggers "recovery" on startup, where everything is truncated/deleted on (prev) LCL. 
A crash between step 3 and 4 will cause core to rename files on startup (well, technically, it will also try to truncate files based on LCL, but it actually won't truncate anything because LCL is up-to-date). Note that on startup there's no insertion into publish queue (we do this before `commit`).

Reply:
ok. I think part of what has me confused here is the interaction between the on-disk publish queue and the checkpoint files. the publish queue is a directory full of numbered .json files now, right? so when you say "queues up checkpoint to publis", that's _also_ an action that involves writing a file and then durably renaming it into place. but it's a different rename. there's a rename that happens before the commit (the one that puts the json file into the publish queue) and then a different set of renames that happen after the commit (the ones that put all the checkpoint xdr files into the publish queue). is that correct?

Reply:
Yes, that's right!